pkg/controller/restore/restore_controller_test.go

```diff
@@ -0,0 +1,272 @@
+	rtc, _, _ := newFakeRestoreController()
+	rtc.enqueueRestore(restore)
+	g.Expect(rtc.queue.Len()).To(Equal(1))
+}
+
+func TestRestoreControllerEnqueueRestoreFailed(t *testing.T) {
+	g := NewGomegaWithT(t)
+	rtc, _, _ := newFakeRestoreController()
+	rtc.enqueueRestore(struct{}{})
+	g.Expect(rtc.queue.Len()).To(Equal(0))
+}
+
+func TestRestoreControllerUpdateRestore(t *testing.T) {
+	g := NewGomegaWithT(t)
+	type testcase struct {
+		name                    string
+		restoreIsInvalid        bool
+		restoreHasBeenCompleted bool
+		restoreHasBeenScheduled bool
+		expectFn                func(*GomegaWithT, *Controller)
+	}
+
+	testFn := func(test *testcase, t *testing.T) {
+		t.Log(test.name)
+
+		restore := newRestore()
+		rtc, _, _ := newFakeRestoreController()
+
+		if test.restoreIsInvalid {
+			restore.Status.Conditions = []v1alpha1.RestoreCondition{
+				{
+					Type:   v1alpha1.RestoreInvalid,
+					Status: corev1.ConditionTrue,
+				},
+			}
+		}
+
+		if test.restoreHasBeenCompleted {
+			restore.Status.Conditions = []v1alpha1.RestoreCondition{
+				{
+					Type:   v1alpha1.RestoreComplete,
+					Status: corev1.ConditionTrue,
+				},
+			}
+		}
+
+		if test.restoreHasBeenScheduled {
+			restore.Status.Conditions = []v1alpha1.RestoreCondition{
+				{
+					Type:   v1alpha1.RestoreScheduled,
+					Status: corev1.ConditionTrue,
+				},
+			}
+		}
+
+		rtc.updateRestore(restore)
+		if test.expectFn != nil {
+			test.expectFn(g, rtc)
+		}
+	}
+
+	tests := []testcase{
+		{
+			name:                    "restore is invalid",
+			restoreIsInvalid:        true,
+			restoreHasBeenCompleted: false,
+			restoreHasBeenScheduled: false,
+			expectFn: func(g *GomegaWithT, rtc *Controller) {
+				g.Expect(rtc.queue.Len()).To(Equal(0))
+			},
+		},
+		{
+			name:                    "restore has been completed",
+			restoreIsInvalid:        false,
+			restoreHasBeenCompleted: true,
+			restoreHasBeenScheduled: false,
+			expectFn: func(g *GomegaWithT, rtc *Controller) {
+				g.Expect(rtc.queue.Len()).To(Equal(0))
+			},
+		},
+		{
+			name:                    "restore has been scheduled",
+			restoreIsInvalid:        false,
+			restoreHasBeenCompleted: false,
+			restoreHasBeenScheduled: true,
+			expectFn: func(g *GomegaWithT, rtc *Controller) {
+				g.Expect(rtc.queue.Len()).To(Equal(0))
+			},
+		},
+		{
+			name:                    "restore is newly created",
+			restoreIsInvalid:        false,
+			restoreHasBeenCompleted: false,
+			restoreHasBeenScheduled: false,
+			expectFn: func(g *GomegaWithT, rtc *Controller) {
+				g.Expect(rtc.queue.Len()).To(Equal(1))
+			},
+		},
+	}
+
+	for i := range tests {
+		testFn(&tests[i], t)
+	}
+}
+
+func TestRestoreControllerSync(t *testing.T) {
+	g := NewGomegaWithT(t)
+	type testcase struct {
+		name                 string
+		addRestoreToIndexer  bool
+		errWhenUpdateRestore bool
+		invalidKeyFn         func(restore *v1alpha1.Restore) string
+		errExpectFn          func(*GomegaWithT, error)
+	}
```

Comment:
We recommend writing unit test in following style:

https://github.com/pingcap/tidb-operator/blob/da683864aef544ea6e6fa921c727d52c3d1bcfcd/pkg/autoscaler/autoscaler/util_test.go#L29-L94

there is no need to define testFn and we can use `go test -run <xxx/sub-xxx>` to test a sub case.

FYI: https://github.com/golang/go/wiki/TableDrivenTests#parallel-testing